src/core.rs

```diff
@@ -0,0 +1,754 @@
+                break;
+            }
+        }
+
+        let atoms = result_rule
+            .body
+            .atoms
+            .into_iter()
+            .filter_map(|atom| match atom.head {
+                HeadOrEq::Eq => {
+                    assert_eq!(atom.args.len(), 2);
+                    match (&atom.args[0], &atom.args[1]) {
+                        (GenericAtomTerm::Var(v1), GenericAtomTerm::Var(v2)) => {
+                            assert_eq!(v1, v2);
+                            None
+                        }
+                        (GenericAtomTerm::Var(_), _) | (_, GenericAtomTerm::Var(_)) => {
+                            panic!("equalities between variable and non-variable arguments should have been canonicalized")
+                        }
+                        (at1, at2) => {
+                            if at1 == at2 {
+                                None
+                            } else {
+                                Some(GenericAtom {
+                                    head: value_eq(&atom.args[0], &atom.args[1]),
+                                    args: vec![
+                                        atom.args[0].clone(),
+                                        atom.args[1].clone(),
+                                        GenericAtomTerm::Literal(Literal::Unit),
+                                    ],
+                                })
+                            }
+                        },
+                    }
+                }
+                HeadOrEq::Symbol(symbol) => Some(GenericAtom {
+                    head: symbol,
+                    args: atom.args,
+                }),
+            })
+            .collect();
+        GenericCoreRule {
+            body: Query { atoms },
+            head: result_rule.head,
+        }
+    }
+}
+
+impl<Head, Leaf> GenericRule<Head, Leaf, ()>
+where
+    Leaf: Clone + Eq + Hash + Debug,
+    Head: Clone,
+{
+    pub(crate) fn to_core_rule(
+        &self,
+        typeinfo: &TypeInfo,
+        mut fresh_gen: impl FreshGen<Head, Leaf>,
+    ) -> Result<GenericCoreRule<HeadOrEq<Head>, Head, Leaf>, TypeError>
+    where
+        Leaf: SymbolLike,
+    {
+        let GenericRule { head, body } = self;
+
+        let (body, _correspondence) = Facts(body.clone()).to_query(typeinfo, &mut fresh_gen);
+        let mut binding = body.get_vars();
+        let (head, _correspondence) =
+            head.to_core_actions(typeinfo, &mut binding, &mut fresh_gen)?;
+        Ok(GenericCoreRule { body, head })
+    }
+
+    fn to_canonicalized_core_rule_impl(
+        &self,
+        typeinfo: &TypeInfo,
+        fresh_gen: impl FreshGen<Head, Leaf>,
+        value_eq: impl Fn(&GenericAtomTerm<Leaf>, &GenericAtomTerm<Leaf>) -> Head,
+    ) -> Result<GenericCoreRule<Head, Head, Leaf>, TypeError>
+    where
+        Leaf: SymbolLike,
+    {
+        let rule = self.to_core_rule(typeinfo, fresh_gen)?;
+        let rule = rule.canonicalize(value_eq);
```

Comment:
nit, optional: I think `Ok(rule.canonicalize(value_eq))` is a little more readable here.